src/extension/infer.rs

```diff
@@ -663,31 +676,98 @@ impl UnificationContext {
         self.results()
     }
 
-    /// Instantiate all variables in the graph with the empty extension set.
+    /// Gather all the transitive dependencies (induced by constraints) of the
+    /// variables in the context.
+    fn search_variable_deps(&self) -> HashSet<Meta> {
+        let mut seen = HashSet::new();
+        let mut new_variables: HashSet<Meta> = self.variables.clone();
+        while !new_variables.is_empty() {
+            let mut newer_variables = HashSet::new();
+            for m in new_variables.into_iter() {
+                if !seen.contains(&m) {
+                    for c in self.get_constraints(&m).unwrap().iter() {
+                        match c {
+                            Constraint::Plus(_, other) => {
+                                newer_variables.insert(self.resolve(*other))
+                            }
+                            Constraint::Equal(other) => {
+                                newer_variables.insert(self.resolve(*other))
+                            }
+                        };
+                    }
+                    seen.insert(m);
+                }
+            }
+            new_variables = newer_variables;
+        }
+        seen
+    }
+
     /// Instantiate all variables in the graph with the empty extension set, or
     /// the smallest solution possible given their constraints.
     /// This is done to solve metas which depend on variables, which allows
     /// us to come up with a fully concrete solution to pass into validation.
+    ///
+    /// Nodes which loop into themselves must be considered as a "minimum" set
+    /// of requirements. If we have
+    ///   1 = 2 + X, ...
+    ///   2 = 1 + x, ...
+    /// then 1 and 2 both definitely contain X, even if we don't know what else.
+    /// So instead of instantiating to the empty set, we'll instantiate to `{X}`
     pub fn instantiate_variables(&mut self) {
-        for m in self.variables.clone().into_iter() {
+        // A directed graph to keep track of `Plus` constraint relationships
+        let mut relations = GraphContainer::<Directed>::new();
+        let mut solutions: HashMap<Meta, ExtensionSet> = HashMap::new();
+
+        let variable_scope = self.search_variable_deps();
+        // If `m` has been merged, [`self.variables`] entry
```

Comment:
Move this comment just inside the loop (the moment when you have just defined `m`)